core/src/main/scala/cats/Monad.scala

```diff
@@ -76,28 +76,37 @@ import syntax.either._
    * Execute an action repeatedly until its result fails to satisfy the given predicate
    * and return that result, discarding all others.
    */
-  def iterateWhile[A](f: F[A])(p: A => Boolean): F[A] = {
+  def iterateWhile[A](f: F[A])(p: A => Boolean): F[A] =
     flatMap(f) { i =>
-      tailRecM(i) { a =>
-        if (p(a))
-          map(f)(_.asLeft[A])
-        else pure(a.asRight[A])
-      }
+      iterateWhileM(i)(_ => f)(p)
     }
-  }
 
   /**
    * Execute an action repeatedly until its result satisfies the given predicate
    * and return that result, discarding all others.
    */
-  def iterateUntil[A](f: F[A])(p: A => Boolean): F[A] = {
+  def iterateUntil[A](f: F[A])(p: A => Boolean): F[A] =
     flatMap(f) { i =>
-      tailRecM(i) { a =>
-        if (p(a))
-          pure(a.asRight[A])
-        else map(f)(_.asLeft[A])
-      }
+      iterateUntilM(i)(_ => f)(p)
     }
-  }
+
+  /**
+    * Apply a monadic function iteratively until its result fails
+    * to satisfy the given predicate and return that result.
+    */
+  def iterateWhileM[A](init: A)(f: A => F[A])(p: A => Boolean): F[A] =
+    tailRecM(init) { a =>
+      if (p(a))
+        map(f(a))(_.asLeft[A])
+      else
+        pure(a.asRight[A])
```

Comment:
Can we replace the `asLeft` and `asRight` with just `Left(..)` and `Right(..)` (in `whileM` as well) ?